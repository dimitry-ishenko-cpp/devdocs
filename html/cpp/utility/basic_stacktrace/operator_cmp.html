    <h1 id="firstHeading" class="firstHeading">operator==, operator&lt;=&gt;<small>(std::basic_stacktrace)</small>
</h1>            <table class="t-dcl-begin"> <tr class="t-dcl t-since-cxx23"> <td> <pre data-language="cpp">template&lt; class Allocator2 &gt;
friend bool operator==( const basic_stacktrace&amp; lhs,
                        const basic_stacktrace&lt;Allocator2&gt;&amp; rhs ) noexcept;</pre>
</td> <td> (1) </td> <td> <span class="t-mark-rev t-since-cxx23">(since C++23)</span> </td> </tr> <tr class="t-dcl t-since-cxx23"> <td> <pre data-language="cpp">template&lt; class Allocator2 &gt;
friend std::strong_ordering
    operator&lt;=&gt;( const basic_stacktrace&amp; lhs,
                 const basic_stacktrace&lt;Allocator2&gt;&amp; rhs ) noexcept;</pre>
</td> <td> (2) </td> <td> <span class="t-mark-rev t-since-cxx23">(since C++23)</span> </td> </tr>  </table> <div class="t-li1">
<span class="t-li">1)</span> Checks if the contents of <code>lhs</code> and <code>rhs</code> are equal, that is, they have the same number of elements and each element in <code>lhs</code> compares equal with the element in <code>rhs</code> at the same position.</div> <div class="t-li1">
 Equivalent to <code><span class="kw1">return</span> <a href="http://en.cppreference.com/w/cpp/algorithm/equal"><span class="kw1577">std::<span class="me2">equal</span></span></a><span class="br0">(</span>lhs.<span class="me1">begin</span><span class="br0">(</span><span class="br0">)</span>, lhs.<span class="me1">end</span><span class="br0">(</span><span class="br0">)</span>, rhs.<span class="me1">begin</span><span class="br0">(</span><span class="br0">)</span>, rhs.<span class="me1">end</span><span class="br0">(</span><span class="br0">)</span><span class="br0">)</span><span class="sy4">;</span></code>.</div> <div class="t-li1">
<span class="t-li">2)</span> Returns the relative order of the numbers of stacktrace entries in <code>lhs</code> and <code>rhs</code> if they are not equal. Otherwise (if the numbers of elements of <code>lhs</code> and <code>rhs</code> are equal), returns the lexicographical order of the elements of <code>lhs</code> and <code>rhs</code>.</div> <div class="t-li1">
 Equivalent to<br><code><span class="kw1">if</span> <span class="br0">(</span><span class="kw4">auto</span> cmp <span class="sy1">=</span> lhs.<span class="me1">size</span><span class="br0">(</span><span class="br0">)</span> <span class="sy1">&lt;=&gt;</span> rhs.<span class="me1">size</span><span class="br0">(</span><span class="br0">)</span><span class="sy4">;</span> cmp <span class="sy3">!</span><span class="sy1">=</span> <span class="nu0">0</span><span class="br0">)</span><br> <p><span class="kw1">return</span> cmp<span class="sy4">;</span><br> <span class="kw1">else</span><br> <span class="kw1">return</span> <a href="http://en.cppreference.com/w/cpp/algorithm/lexicographical_compare_three_way"><span class="kw3240">std::<span class="me2">lexicographical_compare_three_way</span></span></a><span class="br0">(</span>lhs.<span class="me1">begin</span><span class="br0">(</span><span class="br0">)</span>, lhs.<span class="me1">end</span><span class="br0">(</span><span class="br0">)</span>,<br>  rhs.</p>
<span class="me1">begin</span><span class="br0">(</span><span class="br0">)</span>, rhs.<span class="me1">end</span><span class="br0">(</span><span class="br0">)</span><span class="br0">)</span><span class="sy4">;</span></code>.</div> <p>These function templates are not visible to ordinary <a href="../../language/unqualified_lookup.html" title="cpp/language/unqualified lookup">unqualified</a> or <a href="../../language/qualified_lookup.html" title="cpp/language/qualified lookup">qualified lookup</a>, and can only be found by <a href="../../language/adl.html" title="cpp/language/adl">argument-dependent lookup</a> when <code>std::basic_stacktrace&lt;Allocator&gt;</code> is an associated class of the arguments.</p>
<p>The <code>&lt;</code>, <code>&lt;=</code>, <code>&gt;</code>, <code>&gt;=</code>, and <code>!=</code> operators are <a href="../../language/operators.html#Comparison_operators" title="cpp/language/operators">synthesized</a> from operator<span class="sy1">&lt;=&gt;</span> and operator<span class="sy1">==</span> respectively.</p>
<h3 id="Parameters"> Parameters</h3> <table class="t-par-begin"> <tr class="t-par"> <td> lhs, rhs </td> <td> - </td> <td> <code>basic_stacktrace</code>s whose contents to compare </td>
</tr>
</table> <h3 id="Return_value"> Return value</h3> <div class="t-li1">
<span class="t-li">1)</span> <code>true</code> if the contents of <code>lhs</code> and <code>rhs</code> are equal, <code>false</code> otherwise.</div> <div class="t-li1">
<span class="t-li">2)</span> <code>lhs.size() &lt;=&gt; rhs.size()</code> if the result is not <code>std::strong_order::equal</code>, the lexicographical order of the elements of <code>lhs</code> and <code>rhs</code> otherwise.</div> <h3 id="Complexity"> Complexity</h3> <div class="t-li1">
<span class="t-li">1,2)</span> Constant if <code>lhs</code> and <code>rhs</code> are of different size, linear in the size of <code>lhs</code> otherwise.</div> <h3 id="Example"> Example</h3>            <div class="_attribution">
  <p class="_attribution-p">
    &copy; cppreference.com<br>Licensed under the Creative Commons Attribution-ShareAlike Unported License v3.0.<br>
    <a href="https://en.cppreference.com/w/cpp/utility/basic_stacktrace/operator_cmp" class="_attribution-link">https://en.cppreference.com/w/cpp/utility/basic_stacktrace/operator_cmp</a>
  </p>
</div>
