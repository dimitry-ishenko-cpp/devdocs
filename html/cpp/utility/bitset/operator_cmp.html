    <h1 id="firstHeading" class="firstHeading">std::bitset&lt;N&gt;::operator==, std::bitset&lt;N&gt;::operator!=</h1>            <table class="t-dcl-begin">  <tr class="t-dcl-rev-aux"> <td></td> <td rowspan="3">(1)</td> <td></td> </tr> <tr class="t-dcl t-until-cxx11"> <td><pre data-language="cpp">bool operator==( const bitset&amp; rhs ) const;</pre></td>  <td><span class="t-mark-rev t-until-cxx11">(until C++11)</span></td> </tr> <tr class="t-dcl t-since-cxx11"> <td> <pre data-language="cpp">bool operator==( const bitset&amp; rhs ) const noexcept;</pre>
</td>  <td> <span class="t-mark-rev t-since-cxx11">(since C++11)</span> <br><span class="t-mark-rev t-since-cxx23">(constexpr since C++23)</span> </td> </tr>   <tr class="t-dcl-rev-aux"> <td></td> <td rowspan="3">(2)</td> <td></td> </tr> <tr class="t-dcl t-until-cxx11"> <td><pre data-language="cpp">bool operator!=( const bitset&amp; rhs ) const;</pre></td>  <td><span class="t-mark-rev t-until-cxx11">(until C++11)</span></td> </tr> <tr class="t-dcl t-since-cxx11 t-until-cxx20"> <td> <pre data-language="cpp">bool operator!=( const bitset&amp; rhs ) const noexcept;</pre>
</td>  <td> <span class="t-mark-rev t-since-cxx11">(since C++11)</span> <br><span class="t-mark-rev t-until-cxx20">(until C++20)</span> </td> </tr>   </table> <div class="t-li1">
<span class="t-li">1)</span> Returns true if all of the bits in <code>*this</code> and <code>rhs</code> are equal.</div> <div class="t-li1">
<span class="t-li">2)</span> Returns true if any of the bits in <code>*this</code> and <code>rhs</code> are not equal.</div> <table class="t-rev-begin"> <tr class="t-rev t-since-cxx20">
<td> <p>The <code>!=</code> operator is <a href="../../language/default_comparisons.html#Other_defaulted_comparison_operators" title="cpp/language/default comparisons">synthesized</a> from <code>operator==</code>.</p>
</td> <td><span class="t-mark-rev t-since-cxx20">(since C++20)</span></td>
</tr> </table> <h3 id="Parameters"> Parameters</h3> <table class="t-par-begin"> <tr class="t-par"> <td> rhs </td> <td> - </td> <td> bitset to compare </td>
</tr>
</table> <h3 id="Return_value"> Return value</h3> <div class="t-li1">
<span class="t-li">1)</span> <code>true</code> if the value of each bit in <code>*this</code> equals the value of the corresponding bit in <code>rhs</code>, otherwise <code>false</code>.</div> <div class="t-li1">
<span class="t-li">2)</span> <code>true</code> if <code>!(*this == rhs)</code>, otherwise <code>false</code>.</div> <h3 id="Example"> Example</h3> <div class="t-example">
<p>Compare given bitsets to determine if they are identical:</p>
<div class="cpp source-cpp"><pre data-language="cpp">#include &lt;bitset&gt;
#include &lt;iostream&gt;
 
int main()
{
    std::bitset&lt;4&gt; b1(0b0011);
    std::bitset&lt;4&gt; b2(b1);
    std::bitset&lt;4&gt; b3(0b0100);
 
    std::cout &lt;&lt; std::boolalpha;
    std::cout &lt;&lt; "b1 == b2: " &lt;&lt; (b1 == b2) &lt;&lt; '\n';
    std::cout &lt;&lt; "b1 == b3: " &lt;&lt; (b1 == b3) &lt;&lt; '\n';
    std::cout &lt;&lt; "b1 != b3: " &lt;&lt; (b1 != b3) &lt;&lt; '\n';
 
//  b1 == std::bitset&lt;3&gt;{}; // compile-time error: incompatible types
}</pre></div> <p>Output:</p>
<div class="text source-text"><pre data-language="c">b1 == b2: true
b1 == b3: false
b1 != b3: true</pre></div> </div>           <div class="_attribution">
  <p class="_attribution-p">
    &copy; cppreference.com<br>Licensed under the Creative Commons Attribution-ShareAlike Unported License v3.0.<br>
    <a href="https://en.cppreference.com/w/cpp/utility/bitset/operator_cmp" class="_attribution-link">https://en.cppreference.com/w/cpp/utility/bitset/operator_cmp</a>
  </p>
</div>
