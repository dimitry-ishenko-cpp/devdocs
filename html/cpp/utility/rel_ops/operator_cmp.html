    <h1 id="firstHeading" class="firstHeading">std::rel_ops::operator!=,&gt;,&lt;=,&gt;=</h1>            <table class="t-dcl-begin"> <tr class="t-dsc-header"> <th> Defined in header <code><a href="../../header/utility.html" title="cpp/header/utility">&lt;utility&gt;</a></code> </th> <th> </th> <th> </th> </tr> <tr class="t-dcl"> <td><pre data-language="cpp">template&lt; class T &gt;
bool operator!=( const T&amp; lhs, const T&amp; rhs );</pre></td> <td> (1)</td> <td><span class="t-mark">(deprecated in C++20)</span></td> </tr> <tr class="t-dcl"> <td><pre data-language="cpp">template&lt; class T &gt;
bool operator&gt;( const T&amp; lhs, const T&amp; rhs );</pre></td> <td> (2)</td> <td><span class="t-mark">(deprecated in C++20)</span></td> </tr> <tr class="t-dcl"> <td><pre data-language="cpp">template&lt; class T &gt;
bool operator&lt;=( const T&amp; lhs, const T&amp; rhs );</pre></td> <td> (3)</td> <td><span class="t-mark">(deprecated in C++20)</span></td> </tr> <tr class="t-dcl"> <td><pre data-language="cpp">template&lt; class T &gt;
bool operator&gt;=( const T&amp; lhs, const T&amp; rhs );</pre></td> <td> (4)</td> <td><span class="t-mark">(deprecated in C++20)</span></td> </tr>  </table> <p>Given a user-defined <code>operator==</code> and <code>operator&lt;</code> for objects of type <code>T</code>, implements the usual semantics of other comparison operators.</p>
<div class="t-li1">
<span class="t-li">1)</span> Implements <code>operator!=</code> in terms of <code>operator==</code>.</div> <div class="t-li1">
<span class="t-li">2)</span> Implements <code>operator&gt;</code> in terms of <code>operator&lt;</code>.</div> <div class="t-li1">
<span class="t-li">3)</span> Implements <code>operator&lt;=</code> in terms of <code>operator&lt;</code>.</div> <div class="t-li1">
<span class="t-li">4)</span> Implements <code>operator&gt;=</code> in terms of <code>operator&lt;</code>.</div>  <h3 id="Parameters"> Parameters</h3> <table class="t-par-begin"> <tr class="t-par"> <td> lhs </td> <td> - </td> <td> left-hand argument </td>
</tr> <tr class="t-par"> <td> rhs </td> <td> - </td> <td> right-hand argument </td>
</tr>
</table> <h3 id="Return_value"> Return value</h3> <div class="t-li1">
<span class="t-li">1)</span> Returns <code>true</code> if <code>lhs</code> is <i>not equal</i> to <code>rhs</code>.</div> <div class="t-li1">
<span class="t-li">2)</span> Returns <code>true</code> if <code>lhs</code> is <i>greater</i> than <code>rhs</code>.</div> <div class="t-li1">
<span class="t-li">3)</span> Returns <code>true</code> if <code>lhs</code> is <i>less or equal</i> to <code>rhs</code>.</div> <div class="t-li1">
<span class="t-li">4)</span> Returns <code>true</code> if <code>lhs</code> is <i>greater or equal</i> to <code>rhs</code>.</div> <h3 id="Possible_implementation"> Possible implementation</h3> <table class="eq-fun-cpp-table"> <tr> <th>
<a href="#Version_1">(1) <code>operator!=</code></a> </th>
</tr> <tr> <td> <div class="cpp source-cpp"><pre data-language="cpp">namespace rel_ops
{
    template&lt;class T&gt;
    bool operator!=(const T&amp; lhs, const T&amp; rhs)
    {
        return !(lhs == rhs);
    }
}</pre></div> </td>
</tr> <tr> <th>
<a href="#Version_2">(2) <code>operator&gt;</code></a> </th>
</tr> <tr> <td> <div class="cpp source-cpp"><pre data-language="cpp">namespace rel_ops
{
    template&lt;class T&gt;
    bool operator&gt;(const T&amp; lhs, const T&amp; rhs)
    {
        return rhs &lt; lhs;
    }
}</pre></div> </td>
</tr> <tr> <th>
<a href="#Version_3">(3) <code>operator&lt;=</code></a> </th>
</tr> <tr> <td> <div class="cpp source-cpp"><pre data-language="cpp">namespace rel_ops
{
    template&lt;class T&gt;
    bool operator&lt;=(const T&amp; lhs, const T&amp; rhs)
    {
        return !(rhs &lt; lhs);
    }
}</pre></div> </td>
</tr> <tr> <th>
<a href="#Version_4">(4) <code>operator&gt;=</code></a> </th>
</tr> <tr> <td> <div class="cpp source-cpp"><pre data-language="cpp">namespace rel_ops
{
    template&lt;class T&gt;
    bool operator&gt;=(const T&amp; lhs, const T&amp; rhs)
    {
        return !(lhs &lt; rhs);
    }
}</pre></div> </td>
</tr>
</table> <h3 id="Notes"> Notes</h3> <p><a rel="nofollow" class="external text" href="https://www.boost.org/doc/libs/release/libs/utility/operators.htm">Boost.operators</a> provides a more versatile alternative to <code>std::rel_ops</code>.</p>
<p>As of C++20, <code>std::rel_ops</code> are deprecated in favor of <a href="../../language/default_comparisons.html" title="cpp/language/default comparisons"><code>operator&lt;=&gt;</code></a>.</p>
<h3 id="Example"> Example</h3> <div class="t-example"> <div class="cpp source-cpp"><pre data-language="cpp">#include &lt;iostream&gt;
#include &lt;utility&gt;
 
struct Foo
{
    int n;
};
 
bool operator==(const Foo&amp; lhs, const Foo&amp; rhs)
{
    return lhs.n == rhs.n;
}
 
bool operator&lt;(const Foo&amp; lhs, const Foo&amp; rhs)
{
    return lhs.n &lt; rhs.n;
}
 
int main()
{
    Foo f1 = {1};
    Foo f2 = {2};
    using namespace std::rel_ops;
 
    std::cout &lt;&lt; std::boolalpha
              &lt;&lt; "{1} != {2} : " &lt;&lt; (f1 != f2) &lt;&lt; '\n'
              &lt;&lt; "{1} &gt;  {2} : " &lt;&lt; (f1 &gt;  f2) &lt;&lt; '\n'
              &lt;&lt; "{1} &lt;= {2} : " &lt;&lt; (f1 &lt;= f2) &lt;&lt; '\n'
              &lt;&lt; "{1} &gt;= {2} : " &lt;&lt; (f1 &gt;= f2) &lt;&lt; '\n';
}</pre></div> <p>Output:</p>
<div class="text source-text"><pre data-language="c">{1} != {2} : true
{1} &gt;  {2} : false
{1} &lt;= {2} : true
{1} &gt;= {2} : false</pre></div> </div>           <div class="_attribution">
  <p class="_attribution-p">
    &copy; cppreference.com<br>Licensed under the Creative Commons Attribution-ShareAlike Unported License v3.0.<br>
    <a href="https://en.cppreference.com/w/cpp/utility/rel_ops/operator_cmp" class="_attribution-link">https://en.cppreference.com/w/cpp/utility/rel_ops/operator_cmp</a>
  </p>
</div>
