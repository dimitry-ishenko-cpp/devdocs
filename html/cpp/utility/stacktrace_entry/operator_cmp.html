    <h1 id="firstHeading" class="firstHeading">operator==, operator&lt;=&gt;<small>(std::stacktrace_entry)</small>
</h1>            <table class="t-dcl-begin"> <tr class="t-dcl t-since-cxx23"> <td> <pre data-language="cpp">friend constexpr bool operator==( const stacktrace_entry&amp; lhs,
                                  const stacktrace_entry&amp; rhs ) noexcept;</pre>
</td> <td> (1) </td> <td> <span class="t-mark-rev t-since-cxx23">(since C++23)</span> </td> </tr> <tr class="t-dcl t-since-cxx23"> <td> <pre data-language="cpp">friend constexpr std::strong_ordering
    operator&lt;=&gt;( const stacktrace_entry&amp; lhs, const stacktrace_entry&amp; rhs ) noexcept;</pre>
</td> <td> (2) </td> <td> <span class="t-mark-rev t-since-cxx23">(since C++23)</span> </td> </tr>  </table> <div class="t-li1">
<span class="t-li">1)</span> Compares <code>lhs</code> and <code>rhs</code> for equality. Two <code>stacktrace_entry</code> values are equal if and only if they represent the same stacktrace entry, or both of them are empty.</div> <div class="t-li1">
<span class="t-li">2)</span> Gets the relative order between <code>lhs</code> and <code>rhs</code> in the unspecified strict total order over all <code>stacktrace_entry</code> values which is consistent with the equality relation established by <code>operator==</code>.</div> <p>These functions are not visible to ordinary <a href="../../language/unqualified_lookup.html" title="cpp/language/unqualified lookup">unqualified</a> or <a href="../../language/qualified_lookup.html" title="cpp/language/qualified lookup">qualified lookup</a>, and can only be found by <a href="../../language/adl.html" title="cpp/language/adl">argument-dependent lookup</a> when <code>std::stacktrace_entry</code> is an associated class of the arguments.</p>
<p>The <code>&lt;</code>, <code>&lt;=</code>, <code>&gt;</code>, <code>&gt;=</code>, and <code>!=</code> operators are <a href="../../language/operators.html#Comparison_operators" title="cpp/language/operators">synthesized</a> from operator<span class="sy1">&lt;=&gt;</span> and operator<span class="sy1">==</span> respectively.</p>
<h3 id="Parameters"> Parameters</h3> <table class="t-par-begin"> <tr class="t-par"> <td> lhs, rhs </td> <td> - </td> <td> <code>stacktrace_entry</code> values to compare </td>
</tr>
</table> <h3 id="Return_value"> Return value</h3> <div class="t-li1">
<span class="t-li">1)</span> <code>true</code> if two <code>lhs</code> and <code>rhs</code> compare equal, <code>false</code> otherwise.</div> <div class="t-li1">
<span class="t-li">2)</span> <code>std::strong_ordering::equal</code> if <code>lhs</code> and <code>rhs</code> compare equal.</div> <div class="t-li1">
 Otherwise, <code>std::strong_ordering::less</code> if <code>lhs</code> is ordered before <code>rhs</code> in the strict total order.</div> <div class="t-li1">
 Otherwise, <code>std::strong_ordering::greater</code> (in which case <code>rhs</code> is ordered before <code>lhs</code> in the strict total order).</div> <h3 id="Example"> Example</h3>            <div class="_attribution">
  <p class="_attribution-p">
    &copy; cppreference.com<br>Licensed under the Creative Commons Attribution-ShareAlike Unported License v3.0.<br>
    <a href="https://en.cppreference.com/w/cpp/utility/stacktrace_entry/operator_cmp" class="_attribution-link">https://en.cppreference.com/w/cpp/utility/stacktrace_entry/operator_cmp</a>
  </p>
</div>
