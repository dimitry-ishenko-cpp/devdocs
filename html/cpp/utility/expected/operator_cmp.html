    <h1 id="firstHeading" class="firstHeading">operator==<small>(std::expected)</small>
</h1>            <table class="t-dcl-begin"> <tr class="t-dcl t-since-cxx23"> <td> <pre data-language="cpp">template&lt; class T2, class E2 &gt;
  requires (!std::is_void_v&lt;T2&gt;)
friend constexpr bool operator==( const expected&amp; lhs,
                                  const std::expected&lt;T2, E2&gt;&amp; rhs );</pre>
</td> <td> (1) </td> <td> <span class="t-mark-rev t-since-cxx23">(since C++23)</span> <br><span class="t-mark">(T is not <i>cv</i> void)</span> </td> </tr> <tr class="t-dcl t-since-cxx23"> <td> <pre data-language="cpp">template&lt; class T2, class E2 &gt;
  requires std::is_void_v&lt;T2&gt;
friend constexpr bool operator==( const expected&amp; lhs,
                                  const std::expected&lt;T2, E2&gt;&amp; rhs );</pre>
</td> <td> (2) </td> <td> <span class="t-mark-rev t-since-cxx23">(since C++23)</span> <br><span class="t-mark">(T is <i>cv</i> void)</span> </td> </tr> <tr class="t-dcl t-since-cxx23"> <td> <pre data-language="cpp">template&lt; class T2 &gt;
friend constexpr bool operator==( const expected&amp; x, const T2&amp; val );</pre>
</td> <td> (3) </td> <td> <span class="t-mark-rev t-since-cxx23">(since C++23)</span> <br><span class="t-mark">(T is not <i>cv</i> void)</span> </td> </tr> <tr class="t-dcl t-since-cxx23"> <td> <pre data-language="cpp">template&lt; class E2 &gt;
friend constexpr bool operator==( const expected&amp; x,
                                  const unexpected&lt;E2&gt;&amp; e );</pre>
</td> <td> (4) </td> <td> <span class="t-mark-rev t-since-cxx23">(since C++23)</span> </td> </tr>  </table> <p>Performs comparison operations on <code>expected</code> objects.</p>
<span class="t-li">1,2)</span> Compares two <code>expected</code> objects. The objects compare equal if and only if <code>lhs.has_value()</code> and <code>rhs.has_value()</code> are equal, and the contained values are equal. <ul>
<li> For overload <span class="t-v">(1)</span>, if the expressions <code>*lhs == *rhs</code> and <code>lhs.error() == rhs.error()</code> are not well-formed, or if their results are not convertible to <code>bool</code>, the program is ill-formed. </li>
<li> For overload <span class="t-v">(2)</span>, if the expressions <code>lhs.error() == rhs.error()</code> is not well-formed, or if its result is not convertible to <code>bool</code>, the program is ill-formed.</li>
</ul> <span class="t-li">3)</span> Compares <code>expected</code> object with a value. The objects compare equal if and only if <code>x</code> contains an expected value, and the contained value is equal to <code>val</code>. <ul><li> If the expression <code>*x == val</code> is not well-formed, or if its result is not convertible to <code>bool</code>, the program is ill-formed.</li></ul> <span class="t-li">4)</span> Compares <code>expected</code> object with an unexpected value. The objects compare equal if and only if <code>x</code> contains an unexpected value, and the contained value is equal to <code>e.error()</code>. <ul><li> If the expression <code>x.error() == e.error()</code> is not well-formed, or if its result is not convertible to <code>bool</code>, the program is ill-formed.</li></ul> <p>These functions are not visible to ordinary <a href="../../language/unqualified_lookup.html" title="cpp/language/unqualified lookup">unqualified</a> or <a href="../../language/qualified_lookup.html" title="cpp/language/qualified lookup">qualified lookup</a>, and can only be found by <a href="../../language/adl.html" title="cpp/language/adl">argument-dependent lookup</a> when <code>std::expected&lt;T, E&gt;</code> is an associated class of the arguments.</p>
<p>The <code>!=</code> operator is <a href="../../language/default_comparisons.html#Other_defaulted_comparison_operators" title="cpp/language/default comparisons">synthesized</a> from <code>operator==</code>.</p>
<h3 id="Parameters"> Parameters</h3> <table class="t-par-begin"> <tr class="t-par"> <td> lhs, rhs, x </td> <td> - </td> <td> <code>expected</code> object to compare </td>
</tr> <tr class="t-par"> <td> val </td> <td> - </td> <td> value to compare to the expected value contained in <code>x</code> </td>
</tr> <tr class="t-par"> <td> e </td> <td> - </td> <td> value to compare to the unexpected value contained in <code>x</code> </td>
</tr>
</table> <h3 id="Return_value"> Return value</h3> <div class="t-li1">
<span class="t-li">1)</span> If <code>lhs.has_value() != rhs.has_value()</code>, returns <code>false</code>. Otherwise, if <code>lhs.has_value()</code> is true, returns <code>*lhs == *rhs</code>. Otherwise, returns <code>lhs.error() == rhs.error()</code>.</div> <div class="t-li1">
<span class="t-li">2)</span> If <code>lhs.has_value() != rhs.has_value()</code>, returns <code>false</code>. Otherwise, if <code>lhs.has_value()</code> is true, returns <code>true</code>. Otherwise, returns <code>lhs.error() == rhs.error()</code>.</div> <div class="t-li1">
<span class="t-li">3)</span> Returns <code>x.has_value() &amp;&amp; static_cast&lt;bool&gt;(*x == val)</code>.</div> <div class="t-li1">
<span class="t-li">4)</span> Returns <code>!x.has_value() &amp;&amp; static_cast&lt;bool&gt;(x.error() == e.error())</code>.</div> <h3 id="Exceptions"> Exceptions</h3> <p>Throws when and what the comparison throws.</p>
<h3 id="Example"> Example</h3>  <h3 id="See_also"> See also</h3> <table class="t-dsc-begin"> <tr class="t-dsc"> <td> <div><a href="unexpected.html" title="cpp/utility/expected/unexpected"> <span class="t-lines"><span>unexpected</span></span></a></div>
<div><span class="t-lines"><span><span class="t-mark-rev t-since-cxx23">(C++23)</span></span></span></div> </td> <td> represented as an unexpected value <br> <span class="t-mark">(class template)</span>  </td>
</tr> </table>           <div class="_attribution">
  <p class="_attribution-p">
    &copy; cppreference.com<br>Licensed under the Creative Commons Attribution-ShareAlike Unported License v3.0.<br>
    <a href="https://en.cppreference.com/w/cpp/utility/expected/operator_cmp" class="_attribution-link">https://en.cppreference.com/w/cpp/utility/expected/operator_cmp</a>
  </p>
</div>
